Unify cycle detection logic for legacy & LBSE

nikolaszimmermann opened the pull request:
#### 3bd5bd5c81f2cb45f09ff2c8d004241544ce69aa
<pre>
Unify cycle detection logic for legacy &amp; LBSE
<a href="https://bugs.webkit.org/show_bug.cgi?id=268909">https://bugs.webkit.org/show_bug.cgi?id=268909</a>

Reviewed by Rob Buis.

Provide one unique way to prevent recursion in SVG: SVGVisitedRendererTracking.
We already had SVGHitTestCycleDetectionScope to prevent cycles in the legacy
SVG engine during hit-testing, and custom code in SVGSMILElement that provided
the same functionality, but SVGElement, not RenderElement based.

Unify the logic in SVGVisitedRenderingTracking / SVGVisitedElementTracking
and make use of them throughout the legacy SVG engine &amp; LBSE.

It&apos;s used like this:

&lt;example&gt;
bool RenderSVGSomething::someMethod()
{
    static NeverDestroyed&lt;SVGVisitedRendererTracking::VisitedSet&gt; s_visitedSet;

    SVGVisitedRendererTracking recursionTracking(s_visitedSet);
    if (recursionTracking.isVisiting(*this))
        return false;

    SVGVisitedRendererTracking::Scope recursionScope(recursionTracking, *this);

    // We can be sure that if someMethod() is entered again - we&apos;ll gracefully exit as
    // SVGVisitedRendererTracking::isVisiting(const RenderElement&amp;) returns true above.
    performActionThatPotententiallyEntersThisMethodAgain();
    return true;
}
&lt;/example&gt;

Unlike SVGHitTestCycleDetectionScope there is no single static HashSet
shared between all instances of SVGVisitedRendererTracking::Scope - instead
it is passed in as reference to the SVGVisitedRendererTracking constructor.
This allows to seperate the different cylce detection scopes in a fine-grained
way, e.g. per method, per type, etc.

Covered by existing tests.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
Some gardening - cleanup duplicated entries so lint passes again.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::pointInSVGClippingArea const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::drawMarkers):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::applyMaskClipping):
(WebCore::RenderSVGResourceClipper::hitTestClipContent):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::notifyResourceChanged):
(WebCore::RenderSVGResourceContainer::registerResource):
(WebCore::RenderSVGResourceContainer::repaintAllClients const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
(WebCore::RenderSVGResourcePattern::createTileImage const):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::nodeAtFloatPoint):
(WebCore::RenderSVGText::nodeAtPoint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::pointInClippingArea):
(WebCore::SVGHitTestCycleDetectionScope::SVGHitTestCycleDetectionScope): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::~SVGHitTestCycleDetectionScope): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::visitedElements): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::isEmpty): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::isVisiting): Deleted.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGVisitedRendererTracking.h: Added.
(WebCore::SVGVisitedRendererTracking::SVGVisitedRendererTracking):
(WebCore::SVGVisitedRendererTracking::isEmpty const):
(WebCore::SVGVisitedRendererTracking::isVisiting):
(WebCore::SVGVisitedRendererTracking::Scope::Scope):
(WebCore::SVGVisitedRendererTracking::Scope::~Scope):
(WebCore::SVGVisitedRendererTracking::addUnique):
(WebCore::SVGVisitedRendererTracking::removeUnique):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::hitTestClipContent):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::nodeAtPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::nodeAtFloatPoint):
* Source/WebCore/svg/SVGVisitedElementTracking.h: Added.
(WebCore::SVGVisitedElementTracking::SVGVisitedElementTracking):
(WebCore::SVGVisitedElementTracking::isEmpty const):
(WebCore::SVGVisitedElementTracking::isVisiting):
(WebCore::SVGVisitedElementTracking::Scope::Scope):
(WebCore::SVGVisitedElementTracking::Scope::~Scope):
(WebCore::SVGVisitedElementTracking::addUnique):
(WebCore::SVGVisitedElementTracking::removeUnique):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::notifyDependentsIntervalChanged):

Canonical link: <a href="https://commits.webkit.org/274392@main">https://commits.webkit.org/274392@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0b82c5b28013e066b92991a13da5876f1d3e9c54

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38954 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17887 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41307 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41488 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/34671 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20761 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15235 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/32630 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39527 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15070 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/33786 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13084 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13049 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34707 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42765 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35370 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38880 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13766 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11365 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37102 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15372 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8721 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/15033 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/14858 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2850,11 +2850,7 @@ void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const AffineTr
 
-    {
-        // FIXME: Rename SVGHitTestCycleDetectionScope -> SVGResourceCycleDetectionScope
-        SVGHitTestCycleDetectionScope paintingScope(renderer());
```
NOTE: This is removed here, since there shouldn't be scope allocations in random functions -- if not paired with in isVisiting() check. That is fragile and thus should be avoided. Callers need to protect themselves against recursion when they call paintSVGResourceLayer -- this way paintSVGResourceLayer doesn't need to know about the recursion handling itself.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayerModelObject.cpp:
```diff
@@ -579,6 +579,58 @@ RenderSVGResourcePaintServer* RenderLayerModelObject::svgStrokePaintServerResour
     return nullptr;
 }
+
+bool RenderLayerModelObject::pointInSVGClippingArea(const FloatPoint& point) const
```
SVGRenderSupport::pointInClippingArea() explicitly contains cycle detection code - we'd need to add LBSE-only code into SVGRenderSupport to account for that. Since we don't want to add new stuff to mostly-legacy-by-now SVGRenderSupport, the common choice so far was to add stuff to RenderLayerModelObject with a 'SVG' string in the method name - as for applySVGTransform, etc.

That's the reason why this change is included in this PR, it does belong to the cycle detection refactoring.

nikolaszimmermann commented on Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
```diff
@@ -153,18 +154,26 @@ bool RenderSVGContainer::nodeAtPoint(const HitTestRequest& request, HitTestResul
     if (!locationInContainer.intersects(visualOverflowRect))
         return false;
 
+    static NeverDestroyed<SVGVisitedRendererTracking::VisitedSet> s_visitedSet;
```
While this new way of writing is a bit more explicit, and requires the static s_visitedSet definition on the call-site, its way easier to reason about what is happening.

rwlbuis commented on Source/WebCore/rendering/svg/SVGVisitedRendererTracking.h:
```diff
@@ -0,0 +1,82 @@
+
+    class Scope {
+    public:
+        Scope(SVGVisitedRendererTracking& tracking, const RenderElement& renderer, bool condition = true)
```
Is the condition param still used?

nikolaszimmermann replied:
No, but it can indeed be helpful.

rwlbuis replied:
I think it may be better to leave it out until it is needed.

nikolaszimmermann replied:
Ok, no big deal.

rwlbuis commented on Source/WebCore/svg/SVGVisitedElementTracking.h:
```diff
@@ -0,0 +1,82 @@
+
+    class Scope {
+    public:
+        Scope(SVGVisitedElementTracking& tracking, const SVGElement& element, bool condition = true)
```
Ditto?

rwlbuis reviewed (approved):
LGTM.

nikolaszimmermann committed:
Unify cycle detection logic for legacy & LBSE
https://bugs.webkit.org/show_bug.cgi?id=268909

Reviewed by Rob Buis.

Provide one unique way to prevent recursion in SVG: SVGVisitedRendererTracking.
We already had SVGHitTestCycleDetectionScope to prevent cycles in the legacy
SVG engine during hit-testing, and custom code in SVGSMILElement that provided
the same functionality, but SVGElement, not RenderElement based.

Unify the logic in SVGVisitedRenderingTracking / SVGVisitedElementTracking
and make use of them throughout the legacy SVG engine & LBSE.

It's used like this:

<example>
bool RenderSVGSomething::someMethod()
{
    static NeverDestroyed<SVGVisitedRendererTracking::VisitedSet> s_visitedSet;

    SVGVisitedRendererTracking recursionTracking(s_visitedSet);
    if (recursionTracking.isVisiting(*this))
        return false;

    SVGVisitedRendererTracking::Scope recursionScope(recursionTracking, *this);

    // We can be sure that if someMethod() is entered again - we'll gracefully exit as
    // SVGVisitedRendererTracking::isVisiting(const RenderElement&) returns true above.
    performActionThatPotententiallyEntersThisMethodAgain();
    return true;
}
</example>

Unlike SVGHitTestCycleDetectionScope there is no single static HashSet
shared between all instances of SVGVisitedRendererTracking::Scope - instead
it is passed in as reference to the SVGVisitedRendererTracking constructor.
This allows to seperate the different cylce detection scopes in a fine-grained
way, e.g. per method, per type, etc.

Covered by existing tests.

* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
Some gardening - cleanup duplicated entries so lint passes again.
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
* Source/WebCore/rendering/RenderLayerModelObject.cpp:
(WebCore::RenderLayerModelObject::pointInSVGClippingArea const):
* Source/WebCore/rendering/RenderLayerModelObject.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGImage.cpp:
(WebCore::RenderSVGImage::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGPath.cpp:
(WebCore::RenderSVGPath::drawMarkers):
(WebCore::RenderSVGPath::computeMarkerBoundingBox const):
* Source/WebCore/rendering/svg/RenderSVGResourceClipper.cpp:
(WebCore::RenderSVGResourceClipper::applyMaskClipping):
(WebCore::RenderSVGResourceClipper::hitTestClipContent):
(WebCore::RenderSVGResourceClipper::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourceContainer.cpp:
(WebCore::notifyResourceChanged):
(WebCore::RenderSVGResourceContainer::registerResource):
(WebCore::RenderSVGResourceContainer::repaintAllClients const):
* Source/WebCore/rendering/svg/RenderSVGResourceMasker.cpp:
(WebCore::RenderSVGResourceMasker::applyMask):
(WebCore::RenderSVGResourceMasker::resourceBoundingBox):
* Source/WebCore/rendering/svg/RenderSVGResourcePattern.cpp:
(WebCore::RenderSVGResourcePattern::createTileImage const):
* Source/WebCore/rendering/svg/RenderSVGRoot.cpp:
(WebCore::RenderSVGRoot::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGShape.cpp:
(WebCore::RenderSVGShape::nodeAtPoint):
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::nodeAtFloatPoint):
(WebCore::RenderSVGText::nodeAtPoint):
* Source/WebCore/rendering/svg/SVGRenderSupport.cpp:
(WebCore::SVGRenderSupport::pointInClippingArea):
(WebCore::SVGHitTestCycleDetectionScope::SVGHitTestCycleDetectionScope): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::~SVGHitTestCycleDetectionScope): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::visitedElements): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::isEmpty): Deleted.
(WebCore::SVGHitTestCycleDetectionScope::isVisiting): Deleted.
* Source/WebCore/rendering/svg/SVGRenderSupport.h:
* Source/WebCore/rendering/svg/SVGVisitedRendererTracking.h: Added.
(WebCore::SVGVisitedRendererTracking::SVGVisitedRendererTracking):
(WebCore::SVGVisitedRendererTracking::isEmpty const):
(WebCore::SVGVisitedRendererTracking::isVisiting):
(WebCore::SVGVisitedRendererTracking::Scope::Scope):
(WebCore::SVGVisitedRendererTracking::Scope::~Scope):
(WebCore::SVGVisitedRendererTracking::addUnique):
(WebCore::SVGVisitedRendererTracking::removeUnique):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGImage.cpp:
(WebCore::LegacyRenderSVGImage::nodeAtFloatPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResourceClipper.cpp:
(WebCore::LegacyRenderSVGResourceClipper::hitTestClipContent):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGRoot.cpp:
(WebCore::LegacyRenderSVGRoot::nodeAtPoint):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGShape.cpp:
(WebCore::LegacyRenderSVGShape::nodeAtFloatPoint):
* Source/WebCore/svg/SVGVisitedElementTracking.h: Added.
(WebCore::SVGVisitedElementTracking::SVGVisitedElementTracking):
(WebCore::SVGVisitedElementTracking::isEmpty const):
(WebCore::SVGVisitedElementTracking::isVisiting):
(WebCore::SVGVisitedElementTracking::Scope::Scope):
(WebCore::SVGVisitedElementTracking::Scope::~Scope):
(WebCore::SVGVisitedElementTracking::addUnique):
(WebCore::SVGVisitedElementTracking::removeUnique):
* Source/WebCore/svg/animation/SVGSMILElement.cpp:
(WebCore::SVGSMILElement::notifyDependentsIntervalChanged):

Canonical link: https://commits.webkit.org/274392@main
